pkg/sessionctx/variable/session.go

```diff
@@ -1633,35 +1621,38 @@ func (s *SessionVars) GetUseChunkAlloc() bool {
 
 // SetAlloc Attempt to set the buffer pool address
 func (s *SessionVars) SetAlloc(alloc chunk.Allocator) {
-	if !s.EnableReuseCheck {
+	if !s.EnableReuseChunk {
+		s.chunkPool = nil
 		return
 	}
-	s.ChunkPool.Alloc = alloc
+	if alloc == nil {
+		s.chunkPool = nil
+		return
+	}
+	s.chunkPool = chunk.NewReuseHookAllocator(
+		chunk.NewSyncAllocator(alloc),
+		func() {
+			s.StmtCtx.SetUseChunkAlloc()
+		},
+	)
 }
 
-// IsAllocValid check if chunk reuse is enable or ChunkPool is inused.
+// IsAllocValid check if chunk reuse is enable or chunkPool is inused.
 func (s *SessionVars) IsAllocValid() bool {
-	if !s.EnableReuseCheck {
+	if !s.EnableReuseChunk {
 		return false
 	}
-	s.ChunkPool.mu.Lock()
-	defer s.ChunkPool.mu.Unlock()
-	return s.ChunkPool.Alloc != nil
+	return s.chunkPool != nil
 }
 
-// ClearAlloc indicates stop reuse chunk
-func (s *SessionVars) ClearAlloc(alloc *chunk.Allocator, b bool) {
-	if !b {
-		s.ChunkPool.Alloc = nil
+// ClearAlloc indicates stop reuse chunk. If `hasErr` is true, it'll also recreate the `alloc` in parameter.
+func (s *SessionVars) ClearAlloc(alloc *chunk.Allocator, hasErr bool) {
+	if !hasErr {
+		s.chunkPool = nil
 		return
 	}
 
-	// If an error is reported, re-apply for alloc
-	// Prevent the goroutine left before, affecting the execution of the next sql
-	// issuse 38918
-	s.ChunkPool.mu.Lock()
-	s.ChunkPool.Alloc = nil
-	s.ChunkPool.mu.Unlock()
+	s.chunkPool = nil
```

Comment:
Is it safe to remove lock here？

Reply:
The usage of `.chunkPool` field is quite limited. After analyzing the codes, I think it's fine.

I've also discussed with @keeplearning20221 (who is the original author of the re-used chunk). He also agrees it's safe to do so.